### PR TITLE
fix: Make all Promise rejections in asyncData/fetch handled

### PR DESCRIPTION
**Pull request by scryptonite**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
Replaces several calls to `Promise.all` with a wrapper that does the same thing as before, but now accumulates the rejection errors from each of the promises and throws them as an array of errors after all of the supplied promises have resolved or rejected.

This change leaves it to application authors who use `nuxt.renderRoute(path, { req, res })`, `lib/core/middleware/nuxt`, etc... to choose to handle any promise errors that occur inside of a page Component's `asyncData` or `fetch` methods. Before this change, calls to `asyncData` or `fetch` that resulted in a promise rejection or error that would be considered unhandled, **regardless of whether or not there was only one rejection in the promises provided to** `Promise.all`; an `"unhandledRejection"` would be emitted at the process level.

```js
// Results in a caught rejection and also in an "unhandledRejection" event:

await Promise.all([ Promise.reject(12345) ]).catch((error) => error) 
// => 12345 (uncaught: 12345)

await Promise.all([ Promise.reject(1), Promise.reject(2) ]).catch((error) => error) 
// => 1 (uncaught: 1) (uncaught: 2)
```
```js
import { allPromisesHandled } from "lib/app/utils";
// Results in a caught rejection and no "unhandledRejection" event:

await allPromisesHandled([ Promise.reject(12345) ]).catch((errors) => errors) 
// => [ 12345 ]

await allPromisesHandled([ Promise.reject(1), Promise.reject(2) ]).catch((errors) => errors) 
// => [ 1, 2 ]
```
&mdash; [*See `allPromisesHandled` in `lib/app/utils.js`*](https://github.com/scryptonite/nuxt.js/blob/08541707a9f9327091d7980c5c5f2f8fcbacdc08/lib/app/utils.js#L46-L81)

In my application(s), I listen for the process `"unhandledRejection"` event and log the error/promise before shutting down. I wanted to be able to properly handle the promise rejection and also leave my listener for `"unhandledRejection"` in case somewhere in my codebase creates a promise without anticipating and handling its rejection.

Possible side-effects that I've thought of (but I don't think they'll be an issue):
  - A render waiting on a page Component with both an `asyncData` and `fetch`, where one rejects immediately(or at least before the other) will take longer than it did before. `Promise.all` could be thought of as a *'race to any rejection or for all to resolve'*, where `allPromiseHandled` will *'wait for all to reject or resolve'* before resolving or rejecting itself. (I.e. *eager vs patient rejection*)
  - The error thrown by a rejected promise collection is now sometimes an `Array` (if there is only one error involved, that is thrown instead of an array for asyncData/fetch rejections), which may be confusing to some theoretical existing error handling code outside of nuxt's codebase. In some places, I've made the resulting array itself error-like, using a function I made function named [`compactMultipleErrors` in `lib/app/utils.js`](https://github.com/scryptonite/nuxt.js/blob/08541707a9f9327091d7980c5c5f2f8fcbacdc08/lib/app/utils.js#L22-L44); It assigns a `message` and `stack` property to the array. (`message`: based on concatenations of the error messages within, `stack`: first truthy `stack` property in the errors array)

## Checklist:
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [x] I have added tests to cover my changes (if not applicable, please state why)
  - Funny enough, I ran into a jest-related issue while writing a test to cover this change (as in, the test doesn't work properly with a current version of jest: `23.6.0`). See facebook/jest#5620
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by scryptonite** [comment]
I can't seem to determine for sure that the two CI checks this PR fails were due to my fault or not (AFAICT they don't appear to be a result of my commits). If they are, I'd be happy to try to work on getting it resolved.

File: lib/app/client.js [hunk 1]
```diff
@@ -384,8 +386,30 @@ async function render (to, from, next) {
           promises.push(p)
       }
 
-      return Promise.all(promises)
-    }))
+      return allPromisesHandled(promises).catch(async (errors) =>
```
Comment (pi0) [inline lib/app/client.js]: No need to async function. We can Simply `return Promise.resolve/reject` to chain them.

File: lib/app/client.js [hunk 2]
```diff
@@ -384,8 +386,30 @@ async function render (to, from, next) {
           promises.push(p)
       }
 
-      return Promise.all(promises)
-    }))
+      return allPromisesHandled(promises).catch(async (errors) => 
+        Promise.reject(
+          errors.length === 1 ?
```
Comment (pi0) [inline lib/app/client.js]: The logic inside Promise.reject is repeated multiply times (with minor argument differences.) I suggest refactor them into `compactMultiplyErrors`

File: lib/app/utils.js [hunk 3]
```diff
@@ -19,6 +19,67 @@ export function globalHandleError(error) {
   }
 }
 
+// Takes an array of errors and assigns a message and stack to it to make it error-like:
+export function compactMultipleErrors(errors, messagePrefix = 'Multiple Errors', errorJoiner = '\n + \n'){
```
Comment (pi0) [inline lib/app/utils.js]: We loose things like error statusCode here. Maybe we can assign a default value of 500 and then override it if any error had more precise one

File: lib/app/utils.js [hunk 4]
```diff
@@ -19,6 +19,67 @@ export function globalHandleError(error) {
...
+    throw new TypeError('Supplied promises is not an Array')
+  }
+
+  // Array to collect promise errors
+  const errors = []
+
+  // Reduce all promises to a chain of promises resulting in an array of all promise results.
+  return await promises.reduce((chain, promise) => 
+    // With the chain of previous promise results,
+    chain.then(async results => {
```
Comment (pi0) [inline lib/app/utils.js]: No need to async keyword here. Simply return `Promise.resolve` to chain

File: lib/app/utils.js [hunk 5]
```diff
@@ -19,6 +19,67 @@ export function globalHandleError(error) {
...
+        })
+      )
+      return results
+    }),
+    // Initial promise chain is empty:
+    Promise.resolve([])
+  ).then(results => 
+    // If there was any errors, reject with them.
+    errors.length > 0 
+      ? Promise.reject(compactMultipleErrors(errors))
```
Comment (pi0) [inline lib/app/utils.js]: Don't we call compactMultipleErros after it again?

**Review by pi0 — COMMENTED** [review]
Thanks for your time and initial works on this. Handling server side erros is necessary. But we should take more care to prevent unwanted side effects.

**Comment by pi0** [comment]
Another side effect is that error reporters like Sentry may be confused by seeing concatinated errors and unable to log proper traces.

File: lib/app/client.js (hunk 1 above)
Comment (scryptonite) [reply lib/app/client.js]: That's true, I'll remove it in a new commit.

File: lib/app/client.js (hunk 2 above)
Comment (scryptonite) [reply lib/app/client.js]: That's something I can change. I just initially didn't want `compactMultipleErrors` to make the assumption that the author always wanted it to return the first element in the array if it was the only one; plus I didn't think I repeated it enough to warrant DRY-ing it. I'll refactor, if you still want me to, in a new commit.

File: lib/app/utils.js (hunk 3 above)
Comment (scryptonite) [reply lib/app/utils.js]: Is there a particular test case or project that is currently throwing an error that contains a `statusCode` property in something other than the `asyncData` or `fetch` methods? Currently, for asyncData rejections, the first (and likely only) error is thrown directly instead. See [`lib/app/server.js:216`](https://github.com/nuxt/nuxt.js/blob/a3ad155b9a595bd92145dbbdf4ca8661baa175dd/lib/app/server.js#L216)&[`:229`](https://github.com/nuxt/nuxt.js/blob/a3ad155b9a595bd92145dbbdf4ca8661baa175dd/lib/app/server.js#L229), [`lib/app/client.js:391`](https://github.com/nuxt/nuxt.js/blob/a3ad155b9a595bd92145dbbdf4ca8661baa175dd/lib/app/client.js#L391)&[`:404`](https://github.com/nuxt/nuxt.js/blob/a3ad155b9a595bd92145dbbdf4ca8661baa175dd/lib/app/client.js#L404).

If you think it makes sense to add a `statusCode` property to the errors array, should it be based on the first truthy `statusCode` property in the errors array (like already done with the `stack` property), or by some other logic (with the default being 500)? I will refactor it assuming the former in a new commit

File: lib/app/utils.js (hunk 4 above)
Comment (scryptonite) [reply lib/app/utils.js]: Sounds good to me. I will refactor it in a new commit.

**Pull request closed without merging**